### PR TITLE
Network altair rpc

### PR DESCRIPTION
**Pull request by pawanjay176**
## Proposed Changes

Implements rpc changes for altair hard fork.

### CODE REVIEW

**Comment by pawanjay176** [comment]
@AgeManning this is ready for review. All the existing rpc tests are passing.

I'm holding off on adding more tests for post-fork behaviour and compatibility with `V1 rpc` version till we get this PR merged.

Some questions:
1. Should we add functionality to unregister V1 versions at fork boundary? I don't think this is necessary as the spec doesn't strictly enforce doing that. 
> The v1 method **MAY** be unregistered at the fork boundary.
2. Should there be functionality in rpc to completely disable altair (i.e. no V2 version or context bytes)? @michaelsproul Are we using the "Altair never happens" scenario somewhere on the consensus side? https://github.com/sigp/lighthouse/blob/altair/consensus/types/src/chain_spec.rs#L131-L132

**Comment by michaelsproul** [comment]
> Are we using the "Altair never happens" scenario somewhere on the consensus side?

Only for tests. To run the phase0 tests from the EF we hardcode the Altair fork slot to `None` (annoyingly, they test at slot `INT_MAX`). I think it would be fine to error on the network side if the fork slot is `None`, and I can't see any issues with supporting v2 RPC "early".

**Comment by AgeManning** [comment]
I think its fine to not deregister v1.  We can do it in a later update. 

Its probably fine to keep altair non-optional at the moment.

File: beacon_node/eth2_libp2p/src/rpc/codec/ssz_snappy.rs [hunk 1]
```diff
@@ -246,6 +293,14 @@ impl<TSpec: EthSpec> Decoder for SSZSnappyOutboundCodec<TSpec> {
     type Error = RPCError;
 
     fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        // Read the context bytes if required
+        if self.protocol.version == Version::V2 && self.context_bytes.is_none() {
+            let context_bytes = src.split_to(4);
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/codec/ssz_snappy.rs]: This is very dangerous. It will panic if src is < 4 bytes. Are we 100% sure src must be at least 4 bytes long here?

File: beacon_node/eth2_libp2p/src/rpc/protocol.rs [hunk 2]
```diff
@@ -230,12 +240,16 @@ impl ProtocolId {
...
             ),
             Protocol::Goodbye => RpcLimits::new(0, 0), // Goodbye request has no response
-            Protocol::BlocksByRange => {
-                RpcLimits::new(*SIGNED_BEACON_BLOCK_MIN, *SIGNED_BEACON_BLOCK_MAX)
-            }
-            Protocol::BlocksByRoot => {
-                RpcLimits::new(*SIGNED_BEACON_BLOCK_MIN, *SIGNED_BEACON_BLOCK_MAX)
-            }
+            Protocol::BlocksByRange => match self.version {
+                // TODO: use correct altair and base limits
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/protocol.rs]: Is this still supposed to be here?

File: beacon_node/eth2_libp2p/src/rpc/protocol.rs [hunk 3]
```diff
@@ -230,12 +240,16 @@ impl ProtocolId {
...
+                Version::V1 => RpcLimits::new(*SIGNED_BEACON_BLOCK_MIN, *SIGNED_BEACON_BLOCK_MAX),
+                Version::V2 => RpcLimits::new(*SIGNED_BEACON_BLOCK_MIN, *SIGNED_BEACON_BLOCK_MAX),
+            },
+            Protocol::BlocksByRoot => match self.version {
+                // TODO: use correct altair and base limits
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/protocol.rs]: And this?

**Review by AgeManning — COMMENTED** [review]
I dont understand the purpose of the RpcRequestContainer? 

Why do we need to pass around the fork context everywhere, when the RPC and the codecs both already know about it?

**Comment by pawanjay176** [comment]
> I dont understand the purpose of the RpcRequestContainer?

When upgrading the outbound substream here, we need to pass the `fork_context` to the outbound codec.
https://github.com/sigp/lighthouse/blob/f2f6fbbae5c247ad8f050ee4621f468367536dd0/beacon_node/eth2_libp2p/src/rpc/protocol.rs#L504-L514

Just the `RpcRequest` object wouldn't be sufficient here so created a wrapper which also contains the `ForkContext`.

File: beacon_node/eth2_libp2p/src/rpc/codec/ssz_snappy.rs (hunk 1 above)
Comment (pawanjay176) [reply beacon_node/eth2_libp2p/src/rpc/codec/ssz_snappy.rs]: Thank you! will fix this.

File: beacon_node/eth2_libp2p/src/rpc/protocol.rs (hunk 2 above)
Comment (pawanjay176) [reply beacon_node/eth2_libp2p/src/rpc/protocol.rs]: The altair and base beacon blocks have different sizes, so need to use different limits.
Will add those in.

File: beacon_node/eth2_libp2p/src/rpc/protocol.rs (hunk 3 above)
Comment (pawanjay176) [reply beacon_node/eth2_libp2p/src/rpc/protocol.rs]: same as above.

**Comment by AgeManning** [comment]
Yep, as we discussed, I think we should localise the container to the rpc handler.

File: beacon_node/eth2_libp2p/src/behaviour/mod.rs [hunk 4]
```diff
@@ -225,7 +220,7 @@ impl<TSpec: EthSpec> Behaviour<TSpec> {
             .expect("Valid score params and thresholds");
 
         Ok(Behaviour {
-            eth2_rpc: RPC::new(log.clone()),
+            eth2_rpc: RPC::new(fork_context.clone(), log.clone()),
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/behaviour/mod.rs]: Does this need to be cloned?

File: beacon_node/eth2_libp2p/src/rpc/codec/ssz_snappy.rs [hunk 5]
```diff
@@ -131,42 +161,51 @@ impl<TSpec: EthSpec> Decoder for SSZSnappyInboundCodec<TSpec> {
...
+                            // Request type doesn't change, only response type
+                            Protocol::BlocksByRange => Ok(Some(RPCRequest::BlocksByRange(
+                                BlocksByRangeRequest::from_ssz_bytes(&decoded_buffer)?,
+                            ))),
+                            Protocol::BlocksByRoot => {
+                                Ok(Some(RPCRequest::BlocksByRoot(BlocksByRootRequest {
+                                    block_roots: VariableList::from_ssz_bytes(&decoded_buffer)?,
+                                })))
+                            }
+                            _ => Err(RPCError::ErrorResponse(
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/codec/ssz_snappy.rs:215]: There's no v2 Requests for all the other RPC types?
I thought we were going to remove V1 support soon after, this would be we can't accept RPC requests except from BlocksByRoot or BlocksByRange?

File: beacon_node/eth2_libp2p/src/rpc/mod.rs [hunk 6]
```diff
@@ -59,7 +60,7 @@ pub enum RPCReceived<T: EthSpec> {
     ///
     /// The `SubstreamId` is given by the `RPCHandler` as it identifies this request with the
     /// *inbound* substream over which it is managed.
-    Request(SubstreamId, RPCRequest<T>),
+    Request(SubstreamId, RpcRequestContainer<T>),
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/mod.rs]: I thought we were keeping the RequestContainer isolated to the handler. Why is this here?

File: beacon_node/eth2_libp2p/src/rpc/protocol.rs [hunk 7]
```diff
@@ -308,7 +360,13 @@ where
 
             // MetaData requests should be empty, return the stream
             match protocol_name {
-                Protocol::MetaData => Ok((RPCRequest::MetaData(PhantomData), socket)),
+                Protocol::MetaData => Ok((
+                    RpcRequestContainer {
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/protocol.rs]: Can't the return type here just be RpcRequest? 

We don't need the container here right?

File: beacon_node/eth2_libp2p/src/rpc/protocol.rs [hunk 8]
```diff
@@ -333,6 +397,26 @@ where
 // Combines all the RPC requests into a single enum to implement `UpgradeInfo` and
 // `OutboundUpgrade`
 
+#[derive(Debug, Clone)]
+pub struct RpcRequestContainer<TSpec: EthSpec> {
+    pub req: RPCRequest<TSpec>,
+    pub fork_context: Arc<ForkContext>,
+}
+
+impl<T: EthSpec> std::ops::Deref for RpcRequestContainer<T> {
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/protocol.rs]: I think we probably shouldn't do this. Its harder to track down where this is being used. 

We should only use this in handler.rs and protocol.rs and the encoders. It should be hidden from everything else, so probably don't need to deref it.

File: beacon_node/eth2_libp2p/src/rpc/protocol.rs [hunk 9]
```diff
@@ -569,6 +656,12 @@ impl<TSpec: EthSpec> std::fmt::Display for RPCRequest<TSpec> {
     }
 }
 
+impl<TSpec: EthSpec> std::fmt::Display for RpcRequestContainer<TSpec> {
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/protocol.rs]: I dont think we need this either. It should just be an internal container. If we ever need to print it, we should print the whole thing and not pretend like its just an RpcRequest

File: beacon_node/eth2_libp2p/src/rpc/rate_limiter.rs [hunk 10]
```diff
@@ -185,7 +185,7 @@ impl RPCRateLimiter {
     pub fn allows<T: EthSpec>(
         &mut self,
         peer_id: &PeerId,
-        request: &RPCRequest<T>,
+        request: &RpcRequestContainer<T>,
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/rpc/rate_limiter.rs]: Does the rate limiter really need to know about the container?. 
We send a request to the behaviour, the behaviour sends it to the handler, the handler upgrades it to a container and creates streams, once it gets a response we return the response. 
IIRC the rate limiter occurs at the behaviour level.

File: beacon_node/network/src/service.rs [hunk 11]
```diff
@@ -159,13 +159,21 @@ impl<T: BeaconChainTypes> NetworkService<T> {
         // keep track of when our fork_id needs to be updated
         let next_fork_update = next_fork_delay(&beacon_chain);
 
+        // Create a fork context for the given config and genesis validators root
+        let fork_context = Arc::new(ForkContext::new(
+            // TODO(pawan): handle error
+            beacon_chain.slot().unwrap(),
```
Comment (AgeManning) [inline beacon_node/network/src/service.rs]: Need to gracefully handle this error!

**Review by AgeManning — CHANGES_REQUESTED** [review]
The tests for this PR are not passing. Along with these comments, need to fix the tests.

File: beacon_node/eth2_libp2p/src/behaviour/mod.rs [hunk 12]
```diff
@@ -755,7 +748,7 @@ impl<TSpec: EthSpec> Behaviour<TSpec> {
             }
             Ok(RPCReceived::Request(id, request)) => {
                 let peer_request_id = (handler_id, id);
-                match request {
+                match request.req {
```
Comment (AgeManning) [inline beacon_node/eth2_libp2p/src/behaviour/mod.rs]: I dont think we need to expose the behaviour to the container here.

File: consensus/types/src/beacon_block.rs [hunk 13]
```diff
@@ -67,97 +67,11 @@ impl<T: EthSpec> BeaconBlock<T> {
 
     /// Return a block where the block has maximum size.
     pub fn full(spec: &ChainSpec) -> BeaconBlock<T> {
```
Comment (michaelsproul) [inline consensus/types/src/beacon_block.rs]: I wonder if `BeaconBlock::full` should be named differently seeing as it's only used for tests and _shouldn't_ be used for determining the size limit on RPC calls (because it will only return an Altair block for Altair-genesis networks). How would you feel about moving this function to `beacon_node/eth2_libp2p/tests/rpc_tests.rs`, or using `BeaconBlock::empty` in that test instead?

File: beacon_node/eth2_libp2p/src/rpc/codec/ssz_snappy.rs (hunk 5 above)
Comment (pawanjay176) [reply beacon_node/eth2_libp2p/src/rpc/codec/ssz_snappy.rs:215]: The spec defines V2 versions only for `BlocksByRange`, `BlocksByRoot` and `MetaData`. I think the V1 support for only these 3 messages should be removed post fork. Rest of the messages still run on V1 as there are no changes required there.

File: beacon_node/eth2_libp2p/src/rpc/mod.rs (hunk 6 above)
Comment (pawanjay176) [reply beacon_node/eth2_libp2p/src/rpc/mod.rs]: Yeah realised now that it's not required. I had implemented the `expected_responses`, `protocol` and `stream_termination` methods on the `RpcRequestContainer` instead of `RPCRequest`. Because of that, I ended up using the container in places where it was not required. 
Thanks for pointing it out. Fixed now :)

File: beacon_node/eth2_libp2p/src/rpc/protocol.rs (hunk 7 above)
Comment (pawanjay176) [reply beacon_node/eth2_libp2p/src/rpc/protocol.rs]: Fixed.

File: beacon_node/eth2_libp2p/src/rpc/protocol.rs (hunk 8 above)
Comment (pawanjay176) [reply beacon_node/eth2_libp2p/src/rpc/protocol.rs]: Agree. Not even using it now after removing the container in most places.

File: beacon_node/eth2_libp2p/src/rpc/rate_limiter.rs (hunk 10 above)
Comment (pawanjay176) [reply beacon_node/eth2_libp2p/src/rpc/rate_limiter.rs]: Fixed.

File: consensus/types/src/beacon_block.rs (hunk 13 above)
Comment (pawanjay176) [reply consensus/types/src/beacon_block.rs]: Agree. Will remove `BeaconBlock::full`. Will use the `full` method on the Base and Altair variants to figure out the rpc size limits.

**Review by AgeManning — APPROVED** [review]
This looks much nicer.
